Implement Compatibility Mode

nickva opened the pull request:
Compatibilty mode enables writing of `triggered` and `error` states to the
replication documents. It is designed for backward compatibility with the
previous replicator behavior.

Compatibility mode only affects writing of states to the documents, the
scheduling and backoff as well as the new _scheduler/docs and
_scheduler/jobs HTTP endpoints are not affected by the mode change.

It is possible to switch compatibility mode on and off by toggling the
 `replicator.compatibility_mode` configuration flag between `true` and
 `false` values. Compatibility mode can be switched online, however that
 entails a full rescan of all the replicator shards and possibly updating a
 lot of documents. When compatibility mode is  toggled back from `true` to
 `false` document state fields not applicable to the new scheduling replicator
are cleared. (So both toggling from false to true and true to false could
result in document writes).

Implementation notes:

 * A main idea is avoiding an update feedback cycle,  that is when
 a document state update triggers a replication job update,  which,
 in turn, generates another state update and so on... To avoid this situation a
 semantic short circuit is used: discarad all document updates which do not
 change replication parameters. This includes changes to the _replicator_* fields
 used during state updates, or custom user fields. To accomplish this shortcut,
 introduce the idea of a "normalized" #rep{} record. That is a replication record
 which only has fields set that make it possible to compare 2 instances of rep
 records.

* To make normalization work, some options and other list-based rep fields have
to be sorted. Since options are already sorted, don't sort them again, however,
doc_ids filter was not sorted, so make sure to sort just that.

* In order to toggle the mode at runtime, it is necessary to restart doc
processor and multi-db changes monitor. That is done manually from a spawned-ed
process. The reason it is not done via `{stop,....}` is to keep the logs clean.
(couch_replicator_db_changes termination type was switched to a timeout instead
of brutal_kill as well to keep the logs cleaner on mode switch).

* Since the scheduling replicator might force documents to wait a while in the
pending queue before running them, make sure to update the `triggered` state not
only when jobs started successfully but also when they are added to the
replicator the first time. The preserves the rough semantic of "triggered means
replicator has noticed my document and turned into a replication job" idea.

rnewson commented on src/couch_replicator_api_wrap.erl:
```diff
@@ -989,3 +990,48 @@ header_value(Key, Headers, Default) ->
+    };
+
+% Normalize an #db{} record such that it can be used for comparison. This
+% means remove things like pids and also sort options / props.
```
an odd comment for this clause.

nickva replied:
K, will remove

rnewson commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -104,14 +112,15 @@ process_change(DbName, {Change}) ->
             ok = process_updated(Id, JsonRepDoc);
         <<"triggered">> ->
-            couch_replicator_docs:remove_state_fields(DbName, DocId),
+            maybe_remove_state_fields(DbName, DocId, compat_mode()),
```
I'd prefer `maybe_remove_state_fields(DbName, DocId)` here and then a `maybe_remove_state_fields/2` that calls compat_mode() then calls `maybe_remove_state_fields/3`.

nickva replied:
Agreed

rnewson commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -154,8 +173,9 @@ start_link() ->
 
 
 init([]) ->
+    ok = config:listen_for_changes(?MODULE, nil),
```
I think we're optimising prematurely here, it would be simpler to call compat_mode() whenever we need to, since that's comparatively infrequently. We will also be removing this switch in the future.

nickva replied:
this was done to re-parse all documents and fix them up, if we don't do that we don't need this

rnewson commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -206,21 +279,64 @@ code_change(_OldVsn, State, _Extra) ->
-    ok = maybe_start_worker(Id),
-    ok.
+    CurrentRep = current_rep(Id),
+    case normalize_rep(CurrentRep) /= normalize_rep(Rep) of
```
consider `case rep_equals(CurrentRep, Rep) of`

nickva replied:
I started with that but it was just a one-liner which did the /= comparison.

rnewson commented on src/couch_replicator_sup.erl:
```diff
@@ -74,7 +74,7 @@ init(_Args) ->
             {couch_replicator_db_changes, start_link, []},
             permanent,
-            brutal_kill,
+            1000,
```
?

nickva replied:
This is done to make restarts less noisy. We don't need to do that anymore, so this goes away.

sagelywizard commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -31,12 +37,14 @@
     get_json_value/3
 ]).
 
+-define(DEFAULT_COMPATIBILITY, false).
```
Don't we want this to be true by default?

rnewson replied:
no. we want false here but `compatibility_mode = true` to be pushed into local.ini to all cloudant's nodes.

nickva committed:
Implement Compatibility Mode

Compatibilty mode enables writing of `triggered` and `error` states to the
replication documents. It is designed for backward compatibility with the
previous replicator behavior.

Compatibility mode only affects writing of states to the documents, the
scheduling and backoff as well as the new _scheduler/docs and
_scheduler/jobs HTTP endpoints are not affected by the mode change.

Implementation notes:

 * A main idea is avoiding an update feedback cycle,  that is when
 a document state update triggers a replication job update,  which,
 in turn, generates another state update and so on... To avoid this situation a
 semantic short circuit is used: discarad all document updates which do not
 change replication parameters. This includes changes to the _replicator_* fields
 used during state updates, or custom user fields. To accomplish this shortcut,
 introduce the idea of a "normalized" #rep{} record. That is a replication record
 which only has fields set that make it possible to compare 2 instances of rep
 records.

* To make normalization work, some options and other list-based rep fields have
to be sorted. Since options are already sorted, don't sort them again, however,
doc_ids filter was not sorted, so make sure to sort just that.

* Since the scheduling replicator might force documents to wait a while in the
pending queue before running them, make sure to update the `triggered` state not
only when jobs started successfully but also when they are added to the
replicator the first time. The preserves the rough semantic of "triggered means
replicator has noticed my document and turned into a replication job" idea.

mikerhodes commented:
Really good to see this, it should make rolling the replicator improvements out massively easier.

Before merging this it'd be good to get some input from someone like @willholley to check that this fully covers the use-cases of the customers so we don't need to revisit.

garrensmith commented:
This is awesome and will help the dashboard team a lot. Thanks @nickva

nickva commented:
@mikerhodes good point Mike. We are talking with Will, he gave us the initial prod to start heading in this direction.

sagelywizard commented on src/couch_replicator_api_wrap.erl:
```diff
@@ -989,3 +990,46 @@ header_value(Key, Headers, Default) ->
+       timeout = 30000,
+       ibrowse_options = [{k2, v2}, {k1, v1}],
+       retries = 10,
+       http_connections = 20
```
Three space indentation here.

nickva replied:
Oops! will fix

sagelywizard commented on src/couch_replicator_api_wrap.erl:
```diff
@@ -989,3 +990,46 @@ header_value(Key, Headers, Default) ->
+       timeout = HttpDb#httpdb.timeout,
+       ibrowse_options = lists:keysort(1, HttpDb#httpdb.ibrowse_options),
+       retries = HttpDb#httpdb.retries,
+       http_connections = HttpDb#httpdb.http_connections
```
Three space indentation here as well.

nickva replied:
Fixing this too

sagelywizard commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -206,21 +217,68 @@ code_change(_OldVsn, State, _Extra) ->
+
+
+-spec compare_replication_records(#rep{}, #rep{}) -> boolean().
+compare_replication_records(Rep1, Rep2) ->
```
This function seems kinda superfluous to me.

nickva replied:
K, will do a direct comparison there.

sagelywizard commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -206,21 +217,68 @@ code_change(_OldVsn, State, _Extra) ->
+
+normalize_rep(#rep{} = Rep)->
+    #rep{
+       source = couch_replicator_api_wrap:normalize_db(Rep#rep.source),
```
Three space indent

sagelywizard commented on src/couch_replicator_docs.erl:
```diff
@@ -72,19 +76,49 @@ update_doc_completed(DbName, DocId, Stats, StartTime) ->
+    } = Rep,
+    StartTimeBin = couch_replicator_utils:iso8601(StartTime),
+    update_rep_doc(DbName, DocId, [
+            {<<"_replication_state">>, <<"triggered">>},
```
Personal preference that this is only indented one level, a la the call to `update_rep_doc` above.

nickva replied:
Good point will do

sagelywizard commented on src/couch_replicator_scheduler_job.erl:
```diff
@@ -486,7 +486,13 @@ format_status(_Opt, [_PDict, State]) ->
 -spec doc_update_triggered(#rep{}) -> ok.
 doc_update_triggered(#rep{db_name = null}) ->
```
What's this clause for? Do we need to update the rep doc here too?

nickva replied:
_replicate (transient) replications they have their db_name set to `null`

sagelywizard commented on src/couch_replicator_docs.erl:
```diff
@@ -72,19 +76,49 @@ update_doc_completed(DbName, DocId, Stats, StartTime) ->
+    BinRepId = case RepId of
+        {Base, Ext} ->
+            iolist_to_binary([Base, Ext]);
+        _Other ->
```
Can `RepId` be something else?

nickva replied:
Yeap it might not have been calculated (say couldn't fetch filter code from source)

nickva committed:
Remove redunant comparison function

nickva committed:
Fix indentation

garrensmith commented:
@nickva I just want to confirm that this will add back `_replication_state_reason`, `_replication_state` and `_replication_state_time` back to the replication doc?

nickva commented:
@garrensmith yes it will. Specifically, it will update document state with all those fields when replication is triggered and when there is an error (the alternative the document would look as "untriggered that don't not have any of the _replication-\* fields in it).

Because of the internal difference exactly when document gets updated with "triggered" state is slightly different, it is not when it is running the first time but when it is added to the scheduler. But that is a close enough of an approximation I think and is roughly equivalent to "the replicator has noticed my replication document" idea.

One more difference is that there will also be a "_replication_start_time" file which will indicate when the replicator has first noticed the replication document. Hopefully that is not a problem for backwards compatibility.

garrensmith commented:
That all sounds good. That should work fine. Thanks Nick

All misspelling thanks to my iPhone.

On Thu, Nov 10, 2016 at 5:17 PM +0200, "Nick Vatamaniuc" notifications@github.com wrote:

@garrensmith yes it will specifically, it will update document state with all those fields when replication is triggered and when there is an error (the alternative the document would look as "untriggered that don't not have any of the _replication-\* fields in it).

Because of the internal difference exactly when document gets updated with "triggered" state is slightly different, it is not when it is running the first time but when it is added to the scheduler. But that is a close enough of an approximation I think and is roughly equivalent to "the replicator has noticed my replication document" idea.

One more difference is that there will also be a "_replication_start_time" file which will indicate when the replicator has first noticed the replication document. Hopefully that is not a problem for backwards compatibility.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

rnewson commented:
+1
